### ISSUE TITLE
clang-format: adjacent template argument list delimiters `>>` cause QualifierAlignment:Right to break syntax

### ISSUE THREAD

**Issue by khuldraeseth** [llvm/llvm-project#56111]
With `QualifierAlignment: Right`, this snippet is formatted incorrectly by clang-format 15.0.0 regardless of `SpacesInAngles`:
```cpp
foo<const Bar<Baz>>()
// turns into:
// foo< Bar<Baz> const >)
```
Similar problems arise when things other than parentheses follow the template argument list, with the first character always deleted. The spaces in the angle brackets only show up when what follows is an open parenthesis with a match or an open square bracket with a match.

Breaking up the `>>` with a space causes correct behavior. `foo<const Bar<Baz> >()` is formatted to `foo<Bar<Baz> const>()`.

**Comment by mkurdej** [comment]
This seems to be cause by the fact that FormatTokenLexer breaks `<<` and `>>` into two `<` / `>` tokens.
However, at least one of the new tokens has length 2 instead of 1.
Then, when moving the qualifier `const` from before `Bar` to after `Baz>>`, it replaces the text of token `>` (i.e. ` with `const >`). This would be correct but the replaced text range is defined using the incorrect length of 2, so not only `>` is replaced but `>(`.

```
foo<const Bar<Baz>>();
    ^~~~~         ^
    removed       inserted after this token
```

To verify, you can see that formatting `foo<const Bar<Baz>> ();` gives correct result as the token that is eaten up is the space (added before the paren).

Unfortunately, that's not the whole story, there are still failures on other tests (all having `<<` or `>>` in them). Investigating...

**Comment by rymiel** [comment]
This should be fixed by cd7ab4b5c1684dcc60de027700177adfa096b98c
